File: backend/services/preselecta_service.py
import os

import httpx

from backend.services.datacredito_auth import (
    DatacreditoAuthService,
)
from backend.utils.loggers import get_logger
from backend.utils.tls import load_certificates


logger = get_logger("datacredito.preselecta")


class PreselectaService:

    def __init__(self) -> None:

        self.auth = DatacreditoAuthService(
            "preselecta"
        )

        self.url = os.getenv(
            "DATACREDITO_PRESELECTA_SERVICE_URL"
        )

    async def decision(self, payload: dict):

        if not self.url:
            raise RuntimeError(
                "Falta la variable "
                "DATACREDITO_PRESELECTA_SERVICE_URL"
            )

        token = await self.auth.get_access_token()

        headers = {
            "access_token": token,
            "client_id": os.getenv(
                "DATACREDITO_PRESELECTA_CLIENT_ID"
            ),
            "client_secret": os.getenv(
                "DATACREDITO_PRESELECTA_CLIENT_SECRET"
            ),
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

        verify = load_certificates()

        timeout = (
            float(
                os.getenv(
                    "DATACREDITO_TIMEOUT",
                    "30000",
                )
            )
            / 1000
        )

        try:

            async with httpx.AsyncClient(
                verify=verify,
                timeout=timeout,
            ) as client:

                response = await client.post(
                    self.url,
                    json=payload,
                    headers=headers,
                )

            if response.status_code in (401, 403):

                logger.warning(
                    "Token de Preselecta rechazado. "
                    "Renovando token."
                )

                self.auth.clear_token()

                token = await self.auth.get_access_token()

                headers["access_token"] = token

                async with httpx.AsyncClient(
                    verify=verify,
                    timeout=timeout,
                ) as client:

                    response = await client.post(
                        self.url,
                        json=payload,
                        headers=headers,
                    )

            response.raise_for_status()

            return {
                "success": True,
                "status": response.status_code,
                "data": response.json(),
            }

        except httpx.HTTPError as error:

            logger.error(
                "Error consultando Preselecta: %s",
                error,
            )

            raise RuntimeError(
                "Error consumiendo servicio Preselecta"
            ) from error

File: backend/legacy/preselecta.service.js
const axios = require("axios");

const config = require("../config/env");
const {
  createHttpsAgent
} = require("../utils/tls");

const {
  DatacreditoAuthService
} = require("./datacredito-auth.service");

const logger = require("../utils/logger");

class PreselectaService {
  constructor() {
    this.auth =
      new DatacreditoAuthService(
        "preselecta"
      );

    this.httpsAgent =
      createHttpsAgent();
  }

  async decision(payload) {
    const token =
      await this.auth.getAccessToken();

    try {
      const response =
        await axios.post(
          config.datacredito
            .preselecta
            .serviceUrl,

          payload,

          {
            timeout:
              config.datacredito.timeout,

            httpsAgent:
              this.httpsAgent,

            headers: {
              access_token:
                token,

              client_id:
                config.datacredito
                  .preselecta
                  .clientId,

              client_secret:
                config.datacredito
                  .preselecta
                  .clientSecret,

              "Content-Type":
                "application/json",

              Accept:
                "application/json"
            }
          }
        );

      logger.info(
        "Consulta Preselecta ejecutada",
        {
          status:
            response.status
        }
      );

      return {
        success: true,
        status: response.status,
        data: response.data
      };
    } catch (err) {
      if (
        err.response?.status === 401 ||
        err.response?.status === 403
      ) {
        this.auth.clearToken();

        logger.info(
          "Token rechazado. Reintentando autenticación."
        );

        const newToken =
          await this.auth.getAccessToken();

        const retryResponse =
          await axios.post(
            config.datacredito
              .preselecta
              .serviceUrl,

            payload,

            {
              timeout:
                config.datacredito
                  .timeout,

              httpsAgent:
                this.httpsAgent,

              headers: {
                access_token:
                  newToken,

                client_id:
                  config.datacredito
                    .preselecta
                    .clientId,

                client_secret:
                  config.datacredito
                    .preselecta
                    .clientSecret,

                "Content-Type":
                  "application/json",

                Accept:
                  "application/json"
              }
            }
          );

        return {
          success: true,
          status:
            retryResponse.status,
          data:
            retryResponse.data
        };
      }

      logger.error(
        "Error consumiendo Preselecta",
        {
          status:
            err.response?.status,

          response:
            err.response?.data,

          message:
            err.message
        }
      );

      throw err;
    }
  }
}

module.exports =
  PreselectaService;